### PR TITLE
compile time benchmarks for AOTDispatcher (inference/training/subclasses)

### PR DESCRIPTION
**Pull request by bdhirsh**
this adds a few compile time benchmarks for some disjoint paths in AOTDispatcher:

(1) inference vs training code paths
(2) "subclasses" vs "no subclasses" codepaths

Also see https://github.com/pytorch/pytorch/pull/136760 for a partitioner benchmark (I'm not sure why ghstack didn't display the stack nicely)

I ran locally, and got these numbers on the 4 paths:
```
collecting compile time instruction count for aotdispatcher_inference_nosubclass_cpu
compile time instruction count for iteration 0 is 11692348671
compile time instruction count for iteration 1 is 3026287204
compile time instruction count for iteration 2 is 3011467318
compile time instruction count for iteration 3 is 3004485935
compile time instruction count for iteration 4 is 3003087410
collecting compile time instruction count for aotdispatcher_training_nosubclass_cpu
compile time instruction count for iteration 0 is 6068003223
compile time instruction count for iteration 1 is 5585418102
compile time instruction count for iteration 2 is 5581856618
compile time instruction count for iteration 3 is 5581651794
compile time instruction count for iteration 4 is 5578742619
collecting compile time instruction count for aotdispatcher_inference_subclass_cpu
compile time instruction count for iteration 0 is 8634984264
compile time instruction count for iteration 1 is 8633467573
compile time instruction count for iteration 2 is 8632182092
compile time instruction count for iteration 3 is 8632056925
compile time instruction count for iteration 4 is 8632543871
collecting compile time instruction count for aotdispatcher_training_subclass_cpu
compile time instruction count for iteration 0 is 14737239311
compile time instruction count for iteration 1 is 14734346427
compile time instruction count for iteration 2 is 14736493730
compile time instruction count for iteration 3 is 14734121272
compile time instruction count for iteration 4 is 14733852882
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #136670
* #136760
* __->__ #136759

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: benchmarks/dynamo/pr_time_benchmarks/benchmarks/aotdispatcher.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        args = [torch.ones(100, requires_grad=self._training, device=self._device) for _ in range(100)]
+        if self._subclass:
+            args = [TwoTensor(x, x.clone().detach().requires_grad_(self._training)) for x in args]
+        self.args = args
+
+    def _prepare(self):
+        torch._dynamo.reset()
+
+    def _work(self):
+        @torch.compile(backend="aot_eager_decomp_partition", fullgraph=True)
```
Comment (laithsakka) [inline benchmarks/dynamo/pr_time_benchmarks/benchmarks/aotdispatcher.py:41]: do we reach any  inductor caches? or do we stop way beyond?
do we need      with fresh_inductor_cache()

File: benchmarks/dynamo/pr_time_benchmarks/benchmarks/aotdispatcher.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            prefix += "_nosubclass"
+        if self._device == "cpu":
+            prefix += "_cpu"
+        return prefix
+
+    def description(self):
+        return "100 inputs, 100 outputs, each input is added once"
+
+    def _prepare_once(self):
+        args = [torch.ones(100, requires_grad=self._training, device=self._device) for _ in range(100)]
```
Comment (laithsakka) [inline benchmarks/dynamo/pr_time_benchmarks/benchmarks/aotdispatcher.py:32]: nit: _args

File: benchmarks/dynamo/pr_time_benchmarks/benchmarks/aotdispatcher.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+from benchmark_base import BenchmarkBase
+
+import torch
+from torch.testing._internal.two_tensor import TwoTensor
+
+class Benchmark(BenchmarkBase):
+    def __init__(self, *, training, subclass):
+        self._training = training
+        self._subclass = subclass
+        self._device = "cpu"
```
Comment (laithsakka) [inline benchmarks/dynamo/pr_time_benchmarks/benchmarks/aotdispatcher.py:12]: we always use cpu,  is there interesting code paths that we would miss for cuda tensors?

**Review by laithsakka — APPROVED** [review]

**Comment by laithsakka** [comment]
looks good for me, added some comments, will be good for someone else with more AOTDispatcher knowledge to take a look also.

File: benchmarks/dynamo/pr_time_benchmarks/benchmarks/aotdispatcher.py (hunk 1 above)
Comment (bdhirsh) [reply benchmarks/dynamo/pr_time_benchmarks/benchmarks/aotdispatcher.py:41]: nope, the `aot_eager_decomp_partition` backend doesn't ever touch inductor so I removed the inductor cache code (although when the AOTDispatcher cache is eventually on by default we'll need to clear that cache here)

File: benchmarks/dynamo/pr_time_benchmarks/benchmarks/aotdispatcher.py (hunk 3 above)
Comment (bdhirsh) [reply benchmarks/dynamo/pr_time_benchmarks/benchmarks/aotdispatcher.py:12]: Overall AOTDispatcher is pretty device generic. There are a few places where we can end up doing things for cpu vs cuda (e.g. a separate decomp in the `AutogradCPU` vs `AutogradCUDA` dispatch keys), but in general I think this is pretty rare / unlikely to affect compile times in a device-specific way.

**Pull request closed without merging**